mr/local_runner.cc

```diff
@@ -351,6 +353,8 @@ void LocalRunner::Impl::End(ShardFileMap* out_files) {
   }
   dest_mgr_->CloseAllHandles(stop_signal_.load(std::memory_order_acquire));
 
+  all_metric_maps_[current_op_->output().name()] = metric_map;
```

Comment:
ok, so basically everything that is not related to the Environment and IO should reside above Runner....

in other words, I would maintain all_metric_maps_ in the pipeline code.
At the end of the pipeline run, I would create the string representing the counters json and pass it to LocalRunner 
to save. LocalRunner can have a "SaveFile" generic method or `util::Source* CreateSource(...)` if you want a more flexible approach. This way, counters map is always computed correctly and is not dependent on IO code which only needs to implement the serialization routine.

Reply:
I moved the code like you asked, can I push?
Ich habe das Code gezogen, wie du bätest. Kann ich drücken?